In [1]:
import albumentations as A
# with open("/home/pranav/DeepLearning/new.txt","w") as wr:
#     wr.write("Completed first cell")

/home/pranav/.local/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/home/pranav/.local/lib/python3.10/site-packages/albumentations/__init__.py:24: UserWarning: A new version of Albumentations is available: 1.4.21 (you have 1.4.20). Upgrade using: pip install -U albumentations. To disable automatic update checks, set the environment variable NO_ALBUMENTATIONS_UPDATE to 1.
  check_for_updates()


In [2]:
import cv2
import random
import albumentations as A
from matplotlib import pyplot as plt
import time
# with open("/home/pranav/DeepLearning/new.txt","a") as wr:
#     wr.write("The second")

In [3]:
transform = A.Compose([
    A.RandomCrop(width=4500, height=3000, p=1.0),
    A.HorizontalFlip(p=1.0),
    A.VerticalFlip(p=1.0),
    A.Rotate(limit=[60, 240], p=1.0, interpolation=cv2.INTER_NEAREST),
    A.RandomBrightnessContrast(brightness_limit=[-0.2, 0.4], contrast_limit=0.2, p=1.0),
    A.OneOf([
        A.CLAHE (clip_limit=2.0, tile_grid_size=(8, 8), p=0.5),
        A.GridDistortion(p=0.5),
        A.OpticalDistortion(distort_limit=1, shift_limit=0.5, interpolation=cv2.INTER_NEAREST, p=0.5),
    ], p=1.0),
], p=1.0)
# with open("/home/pranav/DeepLearning/new.txt","a") as wr:
#     wr.write("The third")

In [4]:
def visualize(image, mask, original_image=None, original_mask=None):
    fontsize = 16

    if original_image is None and original_mask is None:
        f, ax = plt.subplots(2, 1, figsize=(16, 16), squeeze=True)
        f.set_tight_layout(h_pad=5, w_pad=5)

        ax[0].imshow(image)
        ax[1].imshow(mask)
    else:
        f, ax = plt.subplots(2, 2, figsize=(16, 16), squeeze=True)
        plt.tight_layout(pad=0.2, w_pad=1.0, h_pad=0.01)

        ax[0, 0].imshow(original_image)
        ax[0, 0].set_title('Original Image', fontsize=fontsize)

        ax[1, 0].imshow(original_mask)
        ax[1, 0].set_title('Original Mask', fontsize=fontsize)

        ax[0, 1].imshow(image)
        ax[0, 1].set_title('Transformed Image', fontsize=fontsize)

        ax[1, 1].imshow(mask)
        ax[1, 1].set_title('Transformed Mask', fontsize=fontsize)
        
    plt.savefig('sample_augmented_image.png', facecolor= 'w', transparent= False, bbox_inches= 'tight', dpi= 100)
# with open("/home/pranav/DeepLearning/new.txt","a") as wr:
#     wr.write("The fourth")    

In [5]:
# image = cv2.imread("/home/pranav/DeepLearning/semantic_drone_dataset/training_set/images/040.jpg")
# image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
# mask = cv2.imread("/home/pranav/DeepLearning/semantic_drone_dataset/training_set/gt/semantic/label_images/040.png")
# mask = cv2.cvtColor(mask, cv2.COLOR_BGR2RGB
#                    )

# transformed = transform(image=image, mask=mask)
# transformed_image = transformed['image']
# transformed_mask = transformed['mask']

# cv2.imwrite('./image.png',cv2.cvtColor(transformed_image, cv2.COLOR_BGR2RGB))
# cv2.imwrite('./mask.png',cv2.cvtColor(transformed_mask, cv2.COLOR_BGR2RGB))

# visualize(transformed_image, transformed_mask, image, mask)

In [6]:
images_dir = '/home/pranav/DeepLearning/train_images'
masks_dir = '/home/pranav/DeepLearning/train_masks'
# val_images = 'C:/Users/prana/Deep Learning - Segmentation/val_train_images'
# masksval = "C:/Users/prana/Deep Learning - Segmentation/masksvalimages"
# with open("/home/pranav/DeepLearning/new.txt","a") as wr:
#     wr.write("The third")

In [7]:
import os
import numpy as np

In [8]:
file_names = np.sort(os.listdir("/home/pranav/DeepLearning/semantic_drone_dataset/training_set/images")) 
file_names = np.char.split(file_names, '.')
filenames = np.array([])
for i in range(len(file_names)):
    filenames = np.append(filenames, file_names[i][0])
    

In [9]:
print(filenames)
for file in filenames:
    print(file)

['000' '001' '002' '003' '004' '005' '006' '008' '011' '013' '014' '015'
 '016' '018' '019' '021' '022' '023' '026' '028' '031' '035' '038' '040'
 '041' '042' '043' '044' '045' '047' '049' '051' '052' '053' '055' '056'
 '057' '058' '059' '060' '062' '063' '065' '068' '070' '071' '073' '074'
 '075' '077' '078' '079' '080' '081' '083' '086' '088' '089' '092' '095'
 '098' '099' '100' '101' '102' '103' '104' '106' '107' '109' '110' '111'
 '112' '113' '116' '117' '118' '119' '120' '121' '122' '123' '124' '126'
 '128' '130' '133' '134' '135' '136' '137' '138' '139' '140' '141' '145'
 '146' '147' '148' '149' '150' '153' '154' '155' '156' '157' '158' '159'
 '160' '161' '162' '163' '164' '165' '166' '167' '170' '171' '172' '173'
 '174' '175' '176' '177' '178' '179' '180' '181' '182' '185' '186' '188'
 '190' '192' '193' '194' '195' '198' '199' '200' '202' '204' '206' '207'
 '208' '209' '213' '214' '215' '216' '217' '219' '220' '221' '222' '223'
 '225' '226' '228' '229' '230' '232' '233' '234' '2

In [10]:
def augment_dataset(count):
    '''Function for data augmentation
        Input:
            count - total no. of images after augmentation = initial no. of images * count
        Output:
            writes augmented images (input images & segmentation masks) to the working directory
    '''
    i = 0
    for i in range(count):
        for file in filenames:
            img = cv2.imread("/home/pranav/DeepLearning/semantic_drone_dataset/training_set/images/"+file+'.jpg')
            img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
            mask = cv2.imread("/home/pranav/DeepLearning/semantic_drone_dataset/training_set/gt/semantic/label_images/"+file+'.png')
            mask = cv2.cvtColor(mask, cv2.COLOR_BGR2RGB)
            transformed = transform(image=img, mask=mask)
            transformed_image = transformed['image']
            transformed_mask = transformed['mask']

            cv2.imwrite('/home/pranav/DeepLearning/train_images/aug_images/aug_{}_'.format(str(i+1))+file+'.jpg',cv2.cvtColor(transformed_image, cv2.COLOR_BGR2RGB))
            cv2.imwrite('/home/pranav/DeepLearning/train_masks/aug_masks/aug_{}_'.format(str(i+1))+file+'.png',cv2.cvtColor(transformed_mask, cv2.COLOR_BGR2RGB))

In [11]:
# !mkdir aug_images
# !mkdir aug_masks

In [12]:
#augment_dataset(4)

In [13]:
# !pip install keract
# import keract

In [11]:
import keract


2024-11-24 09:39:04.141637: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1732421344.160017  224288 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1732421344.165533  224288 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-11-24 09:39:04.184443: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [12]:
import pickle
import numpy as np
import pandas as pd
from PIL import Image
from IPython.display import SVG
import matplotlib.pyplot as plt
%matplotlib inline
import os, re, sys, random, shutil, cv2

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import backend as K
from tensorflow.keras.models import Model
from tensorflow.keras.regularizers import l2
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from tensorflow.keras import applications, optimizers
from tensorflow.keras.applications.vgg16 import VGG16, preprocess_input
from tensorflow.keras.preprocessing.image import ImageDataGenerator, load_img, img_to_array
from tensorflow.keras.utils import model_to_dot, plot_model
from tensorflow.keras.callbacks import ModelCheckpoint, ReduceLROnPlateau, EarlyStopping, CSVLogger, LearningRateScheduler, TensorBoard
from tensorflow.keras.layers import Input, Lambda, Activation, Conv2D, MaxPooling2D, BatchNormalization, Add, concatenate, Conv2DTranspose

In [13]:
train_images = '/home/pranav/DeepLearning/train_images'
train_masks = '/home/pranav/DeepLearning/train_masks'


In [14]:
file_names = np.sort(os.listdir(train_images+"/aug_images")) 
file_names = np.char.split(file_names, '.')
filenames = np.array([])
for i in range(len(file_names[:5])):
    filenames = np.append(filenames, file_names[i][0])
print(filenames)

['aug_1_000' 'aug_1_001' 'aug_1_002' 'aug_1_003' 'aug_1_004']


In [18]:
# def show_data(files, original_images_dir, label_images_dir):

#     for file in files:
#         fig, axs = plt.subplots(1, 2, figsize=(20, 10), constrained_layout=True)

#         axs[0].imshow(Image.open(original_images_dir+"/aug_images/"+str(file)+'.jpg'))
#         axs[0].set_title('Original Image', fontdict = {'fontsize':20})
#         axs[0].set_xticks(np.arange(0, 6500, 1000))
#         axs[0].set_yticks(np.arange(0, 4500, 1000))
#         axs[0].grid(False)
#         axs[0].axis(True)

#         semantic_label_image = Image.open(label_images_dir+"/aug_masks/"+str(file)+'.png')
#         semantic_label_image = np.asarray(semantic_label_image)
#         axs[1].imshow(semantic_label_image)
#         axs[1].set_title('Semantic Segmentation Mask', fontdict = {'fontsize':20})
#         axs[1].set_xticks(np.arange(0, 6500, 1000))
#         axs[1].set_yticks(np.arange(0, 4500, 1000))
#         axs[1].grid(False)
#         axs[1].axis(True)

#         plt.savefig('image_'+file, facecolor= 'w', transparent= False, bbox_inches= 'tight', dpi= 100)
#         plt.show()
    
# show_data(filenames[:5], train_images, train_masks)    

In [15]:
augmented_files = ['092','118', '228', '277', '376']

def show_augmented_images(files, original_images_dir):
    for file in files:
        fig, axs = plt.subplots(1, 5, figsize=(30, 6), constrained_layout=True)
        for i in range(5):
            if i == 0:
                axs[i].imshow(Image.open(original_images_dir+str(file)+'.jpg'))
                axs[i].set_title('Original Image: {}.jpg'.format(file), fontdict = {'fontsize':20})
                axs[i].set_xticks(np.arange(0, 6500, 1000))
                axs[i].set_yticks(np.arange(0, 4500, 1000))
                axs[i].grid(False)
                axs[i].axis(True)
            else:
                axs[i].imshow(Image.open("C:/Users/prana/Deep Learning - Segmentation/aug_images/"+'aug_'+str(i)+'_'+str(file)+'.jpg'))
                axs[i].set_title('Augmented Image: aug_'+str(i)+'_'+str(file)+'.jpg', fontdict = {'fontsize':20})
                axs[i].set_xticks(np.arange(0, 4500, 1000))
                axs[i].set_yticks(np.arange(0, 3001, 1000))
                axs[i].grid(False)
                axs[i].axis(True)

        plt.savefig('aug_image_'+file, facecolor= 'w', transparent= False, bbox_inches= 'tight', dpi= 100)
        plt.show()
    
def show_augmented_masks(files, label_images_dir):
    for file in files:
        fig, axs = plt.subplots(1, 5, figsize=(30, 6), constrained_layout=True)
        for i in range(5):
            if i == 0:
                axs[i].imshow(Image.open(label_images_dir+str(file)+'.png'))
                axs[i].set_title('Original Mask: {}.png'.format(file), fontdict = {'fontsize':20})
                axs[i].set_xticks(np.arange(0, 6500, 1000))
                axs[i].set_yticks(np.arange(0, 4500, 1000))
                axs[i].grid(False)
                axs[i].axis(True)
            else:
                axs[i].imshow(Image.open("C:/Users/prana/Deep Learning - Segmentation/aug_masks/"+"aug_"+str(i)+'_'+str(file)+'.png'))
                axs[i].set_title('Augmented Mask: aug_'+str(i)+'_'+str(file)+'.png', fontdict = {'fontsize':20})
                axs[i].set_xticks(np.arange(0, 4500, 1000))
                axs[i].set_yticks(np.arange(0, 3001, 1000))
                axs[i].grid(False)
                axs[i].axis(True)

        plt.savefig('aug_mask_'+file, facecolor= 'w', transparent= False, bbox_inches= 'tight', dpi= 100)
        plt.show()

In [20]:
# show_augmented_images(augmented_files, train_images)
# show_augmented_masks(augmented_files, train_masks)
# print(train_images)
# print(train_masks)

In [16]:
class_dict_df = pd.read_csv("/home/pranav/DeepLearning/semantic_drone_dataset/training_set/gt/semantic/class_dict.csv", index_col=False, skipinitialspace=True)
print(class_dict_df["name"])


0       unlabeled
1      paved-area
2            dirt
3           grass
4          gravel
5           water
6           rocks
7            pool
8      vegetation
9            roof
10           wall
11         window
12           door
13          fence
14     fence-pole
15         person
16            dog
17            car
18        bicycle
19           tree
20      bald-tree
21      ar-marker
22       obstacle
23    conflicting
Name: name, dtype: object


In [17]:
label_names= list(class_dict_df.name)
label_codes = []
r= np.asarray(class_dict_df.r)
g= np.asarray(class_dict_df.g)
b= np.asarray(class_dict_df.b)

for i in range(len(class_dict_df)):
    label_codes.append(tuple([r[i], g[i], b[i]]))
    
label_codes[:5], label_names[:5]

([(np.int64(0), np.int64(0), np.int64(0)),
  (np.int64(128), np.int64(64), np.int64(128)),
  (np.int64(130), np.int64(76), np.int64(0)),
  (np.int64(0), np.int64(102), np.int64(0)),
  (np.int64(112), np.int64(103), np.int64(87))],
 ['unlabeled', 'paved-area', 'dirt', 'grass', 'gravel'])

In [18]:
label_codes, label_names

([(np.int64(0), np.int64(0), np.int64(0)),
  (np.int64(128), np.int64(64), np.int64(128)),
  (np.int64(130), np.int64(76), np.int64(0)),
  (np.int64(0), np.int64(102), np.int64(0)),
  (np.int64(112), np.int64(103), np.int64(87)),
  (np.int64(28), np.int64(42), np.int64(168)),
  (np.int64(48), np.int64(41), np.int64(30)),
  (np.int64(0), np.int64(50), np.int64(89)),
  (np.int64(107), np.int64(142), np.int64(35)),
  (np.int64(70), np.int64(70), np.int64(70)),
  (np.int64(102), np.int64(102), np.int64(156)),
  (np.int64(254), np.int64(228), np.int64(12)),
  (np.int64(254), np.int64(148), np.int64(12)),
  (np.int64(190), np.int64(153), np.int64(153)),
  (np.int64(153), np.int64(153), np.int64(153)),
  (np.int64(255), np.int64(22), np.int64(96)),
  (np.int64(102), np.int64(51), np.int64(0)),
  (np.int64(9), np.int64(143), np.int64(150)),
  (np.int64(119), np.int64(11), np.int64(32)),
  (np.int64(51), np.int64(51), np.int64(0)),
  (np.int64(190), np.int64(250), np.int64(190)),
  (np.int64(11

In [19]:
code2id = {v:k for k,v in enumerate(label_codes)}
id2code = {k:v for k,v in enumerate(label_codes)}

name2id = {v:k for k,v in enumerate(label_names)}
id2name = {k:v for k,v in enumerate(label_names)}

In [20]:
def rgb_to_onehot(rgb_image, colormap = id2code):
    '''Function to one hot encode RGB mask labels
        Inputs: 
            rgb_image - image matrix (eg. 256 x 256 x 3 dimension numpy ndarray)
            colormap - dictionary of color to label id
        Output: One hot encoded image of dimensions (height x width x num_classes) where num_classes = len(colormap)
    '''
    num_classes = len(colormap)
    shape = rgb_image.shape[:2]+(num_classes,)
    encoded_image = np.zeros( shape, dtype=np.int8 )
    for i, cls in enumerate(colormap):
        encoded_image[:,:,i] = np.all(rgb_image.reshape( (-1,3) ) == colormap[i], axis=1).reshape(shape[:2])
    return encoded_image


def onehot_to_rgb(onehot, colormap = id2code):
    '''Function to decode encoded mask labels
        Inputs: 
            onehot - one hot encoded image matrix (height x width x num_classes)
            colormap - dictionary of color to label id
        Output: Decoded RGB image (height x width x 3) 
    '''
    single_layer = np.argmax(onehot, axis=-1)
    output = np.zeros( onehot.shape[:2]+(3,) )
    for k in colormap.keys():
        output[single_layer==k] = colormap[k]
    return np.uint8(output)

In [21]:
# Normalizing only frame images, since masks contain label info
data_gen_args = dict(rescale=1./255)
mask_gen_args = dict()

train_frames_datagen = ImageDataGenerator(**data_gen_args)
train_masks_datagen = ImageDataGenerator(**mask_gen_args)
val_frames_datagen = ImageDataGenerator(**data_gen_args)
val_masks_datagen = ImageDataGenerator(**mask_gen_args)
# Seed defined for aligning images and their masks
seed = 1

In [22]:
def TrainAugmentGenerator(train_images_dir, train_masks_dir, seed = 1, batch_sizess = 8, target_sizess = (512, 512)):
    '''Train Image data generator
        Inputs: 
            seed - seed provided to the flow_from_directory function to ensure aligned data flow
            batch_size - number of images to import at a time
            train_images_dir - train images directory
            train_masks_dir - train masks directory
            target_size - tuple of integers (height, width)
            
        Output: Decoded RGB image (height x width x 3) 
    '''
    train_image_generator = train_frames_datagen.flow_from_directory(
   train_images_dir,
    batch_size= batch_sizess, 
    seed = seed, 
    target_size = target_sizess)

    train_mask_generator = train_masks_datagen.flow_from_directory(
    train_masks_dir,
    batch_size = batch_sizess, 
    seed = seed, 
    target_size = target_sizess)

    while True:
        X1i = next(train_image_generator)
        X2i = next(train_mask_generator)
        
        #One hot encoding RGB images
        mask_encoded = [rgb_to_onehot(X2i[0][x,:,:,:], id2code) for x in range(X2i[0].shape[0])]
        if(X1i[0] is None or mask_encoded is None):
            print("I am sending Null object to you")
        yield X1i[0], np.asarray(mask_encoded)

def ValAugmentGenerator(val_images_dr, masksvaldr, seed = 1, batch_sizess = 8, target_sizess = (512, 512)):
    '''Train Image data generator
        Inputs: 
            seed - seed provided to the flow_from_directory function to ensure aligned data flow
            batch_size - number of images to import at a time
            train_images_dir - train images directory
            train_masks_dir - train masks directory
            target_size - tuple of integers (height, width)
            
        Output: Decoded RGB image (height x width x 3) 
    '''
    val_image_generator = val_frames_datagen.flow_from_directory(
   val_images_dr,
    batch_size= batch_sizess, 
    seed = seed, 
    target_size = target_sizess)

    val_mask_generator = val_masks_datagen.flow_from_directory(
    masksvaldr,
    batch_size = batch_sizess, 
    seed = seed, 
    target_size = target_sizess)

    while True:
        X1i = next(val_image_generator)
        X2i = next(val_mask_generator)
        
        #One hot encoding RGB images
        mask_encoded = [rgb_to_onehot(X2i[0][x,:,:,:], id2code) for x in range(X2i[0].shape[0])]
        if(X1i[0] is None or mask_encoded is None):
            print("I am sending Null object to you")
        yield X1i[0], np.asarray(mask_encoded)

        


In [28]:
# !mkdir pretrained_weight

In [23]:
import numpy as np
import os
batch_size = 8
num_train_samples = len(np.sort(os.listdir(train_images)))
steps_per_epoch = np.ceil(float(num_train_samples) / float(batch_size))
print('steps_per_epoch: ', steps_per_epoch)
print(train_images)
print(train_masks)
with open("/home/pranav/DeepLearning/new.txt","a") as wr:
    wr.write("The sixth")

steps_per_epoch:  1.0
/home/pranav/DeepLearning/train_images
/home/pranav/DeepLearning/train_masks


In [2]:
import tensorflow as tf
from tensorflow.keras.applications import EfficientNetB0
from tensorflow.keras import layers, models

In [13]:
from tensorflow.keras import backend as K
import tensorflow as tf
from tensorflow.keras.layers import (
    Input, Conv2D, BatchNormalization, ReLU, DepthwiseConv2D, GlobalAveragePooling2D,
    Dense, Reshape, Multiply
)
from tensorflow.keras.models import Model

# def dice_score(y_true, y_pred, smooth=1e-6):
#     # Convert y_pred from probabilities to class predictions
#      tf.print(y_true_f.shape)
#     tf.print(y_pred_f.shape)
#     y_pred_classes = tf.argmax(y_pred, axis=-1)

#     # Convert y_true to the same type and shape as y_pred_classes
#     y_true = tf.cast(y_true, tf.int64)
#     print(y_true)
#     print(y_pred_classes)
#     # Ensure both y_true and y_pred_classes have the same shape (flatten them)
#     y_true_f = tf.reshape(y_true, [-1])
#     y_pred_f = tf.reshape(y_pred_classes, [-1])

#     # Calculate the Dice score
#     intersection = tf.reduce_sum(tf.cast(y_true_f == y_pred_f, tf.float32))
#     dice = (2. * intersection + smooth) / (tf.reduce_sum(tf.cast(y_true_f != -1, tf.float32)) + 
#                                             tf.reduce_sum(tf.cast(y_pred_f != -1, tf.float32)) + smooth)
#     print(y_true_f.shape)
#     print(y_pred_f.shape)
#     return dice
# def simplemetric(y_true, y_pred):
#     return tf.constant(0.5)  
def dice_coef(y_true, y_pred):
    return (2. * K.sum(y_true * y_pred) + 1.) / (K.sum(y_true) + K.sum(y_pred) + 1.)

# class SimpleMetric(tf.keras.metrics.Metric):
#     def __init__(self, name='simple_constant_metric', **kwargs):
#         super(SimpleConstantMetric, self).__init__(name=name, **kwargs)
#         self.value = self.add_weight(name="value", initializer="zeros")

#     def update_state(self, y_true, y_pred, sample_weight=None):
#         # Update state with the constant value (0.5)
#         self.value.assign(0.5)

#     def result(self):
#         # Return the current value of the metric
#         return self.value

#     def reset_states(self):
#         # Reset metric state
#         self.value.assign(0.0)
class DiceCoefficient(tf.keras.metrics.Metric):
    def __init__(self, name='dice_coefficient', smooth=1e-6, **kwargs):
        super(DiceCoefficient, self).__init__(name=name, **kwargs)
        self.smooth = smooth
        self.intersection = self.add_weight(name='intersection', initializer='zeros')
        self.union = self.add_weight(name='union', initializer='zeros')

    def update_state(self, y_true, y_pred, sample_weight=None):
        # Flatten the tensors to calculate the Dice coefficient
        y_true_f = tf.cast(tf.reshape(y_true, [-1]), tf.float32)
        y_pred_f = tf.cast(tf.reshape(y_pred, [-1]), tf.float32)
        
        # Calculate the intersection (true positives)
        intersection = tf.reduce_sum(y_true_f * y_pred_f)
        
        # Calculate the union (sum of both sets)
        union = tf.reduce_sum(y_true_f) + tf.reduce_sum(y_pred_f)
        
        # Update the total intersection and union state
        self.intersection.assign_add(intersection)
        self.union.assign_add(union)

    def result(self):
        # Return the Dice coefficient
        dice = (2. * self.intersection + self.smooth) / (self.union + self.smooth)
        return dice  # This will return a 'tf.Tensor' (EagerTensor in eager mode)

    def reset_states(self):
        # Reset the intersection and union for the next epoch
        self.intersection.assign(0.0)
        self.union.assign(0.0)
    def get_config(self):  # Renamed from `get_config()` to `get_dice_config()`
        dt = super(DiceCoefficient, self).get_config()
        dt.update({
            "smooth": self.smooth,
        })
        return dt   
class TimeHistory(tf.keras.callbacks.Callback):
    def __init__(self, file_path="epoch_times.txt"):
        super(TimeHistory, self).__init__()
        self.file_path = file_path  # Custom file path for saving the times

    def on_train_begin(self, logs=None):
        self.epoch_times = []  # Initialize a list to store times for each epoch

    def on_epoch_begin(self, epoch, logs=None):
        self.start_time = time.time()  # Record the start time of the epoch

    def on_epoch_end(self, epoch, logs=None):
        epoch_duration = time.time() - self.start_time  # Calculate the duration of the epoch
        self.epoch_times.append(epoch_duration)  # Append to the list

    def on_train_end(self, logs=None):
        # Ensure the directory exists
        os.makedirs(os.path.dirname(self.file_path), exist_ok=True)
        
        # Store the epoch times list in the specified file path
        with open(self.file_path, "w") as file:
            file.write(str(self.epoch_times))
        print(f"Epoch times saved to '{self.file_path}'.")    
import tensorflow as tf
from tensorflow.keras.layers import (
    Input, Conv2D, BatchNormalization, ReLU, DepthwiseConv2D, GlobalAveragePooling2D,
    Dense, Reshape, Multiply, UpSampling2D, Concatenate
)
from tensorflow.keras.models import Model

import tensorflow as tf
from tensorflow.keras.layers import (
    Input, Conv2D, BatchNormalization, ReLU, DepthwiseConv2D, GlobalAveragePooling2D,
    Dense, Reshape, Multiply, UpSampling2D, Concatenate
)
from tensorflow.keras.models import Model
        
def se_block(input_tensor, reduction_ratio=16):
    channel_axis = -1
    filters = input_tensor.shape[channel_axis]
    se = GlobalAveragePooling2D()(input_tensor)
    se = Dense(filters // reduction_ratio, activation='relu')(se)
    se = Dense(filters, activation='sigmoid')(se)
    se = Reshape([1, 1, filters])(se)
    x = Multiply()([input_tensor, se])
    return x        

def bottleneck_block(x, in_channels, out_channels, expansion_factor, stride, use_se=False):
    expanded_channels = in_channels * expansion_factor

    # Expansion phase
    if expansion_factor != 1:
        x = Conv2D(expanded_channels, (1, 1), padding='same', use_bias=False)(x)
        x = BatchNormalization()(x)
        x = tf.keras.layers.ReLU()(x)

    # Depthwise convolution
    x = DepthwiseConv2D((3, 3), strides=stride, padding='same', use_bias=False)(x)
    x = BatchNormalization()(x)
    x = tf.keras.layers.ReLU()(x)

    # Squeeze-and-Excitation (optional)
    if use_se:
        x = se_block(x)

    # Projection phase
    x = Conv2D(out_channels, (1, 1), padding='same', use_bias=False)(x)
    x = BatchNormalization()(x)

    return x

# MobileNetV2 with SE Blocks and Decoder
def mobilenet_v2_with_se_and_decoder(input_shape=(512, 512, 3), num_classes=24):
    inputs = Input(input_shape)

    # Initial convolution layer
    x = Conv2D(32, (3, 3), strides=2, padding='same', use_bias=False)(inputs)
    x = BatchNormalization()(x)
    x = tf.keras.layers.ReLU()(x)

    # Store skip connections for decoder
    skipss = []

    # Bottleneck layers
    bottleneck_settings = [
        # (expansion_factor, out_channels, num_blocks, stride, use_se)
        (1, 16, 1, 1, False),
        (6, 24, 2, 2, False),
        (6, 32, 3, 2, True),
        (6, 64, 4, 2, True),
        (6, 96, 3, 1, True),
        (6, 160, 3, 2, True),
        (6, 320, 1, 1, True),
    ]

    in_channels = 32
    for expansion_factor, out_channels, num_blocks, stride, use_se in bottleneck_settings:
        for i in range(num_blocks):
            s = stride if i == 0 else 1  # Only apply stride on the first block
            x = bottleneck_block(x, in_channels, out_channels, expansion_factor, s, use_se)
            if s > 1:  # Save for skip connection if downsampling
                skipss.append(x)
            in_channels = out_channels

    # Decoder: UpSample and Concatenate
    for skip in reversed(skipss):
        x = UpSampling2D((2, 2))(x)
        
        # Ensure the skip connection matches the shape of x
        skip = UpSampling2D(size=(x.shape[1] // skip.shape[1], x.shape[2] // skip.shape[2]))(skip)
        
        x = Concatenate()([x, skip])  # Concatenate skip connections
        x = Conv2D(skip.shape[-1], (3, 3), padding='same', activation='relu')(x)
        x = BatchNormalization()(x)

    # Final convolution layer to reach the same shape as input
    x = UpSampling2D((2, 2))(x)
    x = Conv2D(num_classes, (1, 1), activation='softmax', padding='same')(x)

    return Model(inputs, x)

In [14]:
vgg16_unet = mobilenet_v3_with_se_and_decoder(num_classes = 24, input_shape = (512, 512, 3), lr_init=0.0001)
vgg16_unet.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer_2 (InputLayer)    │ (None, 512, 512, 3)       │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ conv2d_41 (Conv2D)            │ (None, 256, 256, 32)      │             864 │ input_layer_2[0][0]        │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ batch_normalization_62        │ (None, 256, 256, 32)      │             128 │ conv2d_41[0][0]            │
│ (BatchNormalization)          │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ re_lu_42 (ReLU)               │ (None, 256, 256, 32)      │               0 │ batch_normalization_62[0]… │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ depthwise_conv2d_21           │ (None, 256, 256, 32)      │             288 │ re_lu_42[0][0]             │
│ (DepthwiseConv2D)             │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ batch_normalization_63        │ (None, 256, 256, 32)      │             128 │ depthwise_conv2d_21[0][0]  │
│ (BatchNormalization)          │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ re_lu_43 (ReLU)               │ (None, 256, 256, 32)      │               0 │ batch_normalization_63[0]… │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ conv2d_42 (Conv2D)            │ (None, 256, 256, 16)      │             512 │ re_lu_43[0][0]             │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ batch_normalization_64        │ (None, 256, 256, 16)      │              64 │ conv2d_42[0][0]            │
│ (BatchNormalization)          │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ conv2d_43 (Conv2D)            │ (None, 256, 256, 96)      │           1,536 │ batch_normalization_64[0]… │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ batch_normalization_65        │ (None, 256, 256, 96)      │             384 │ conv2d_43[0][0]            │
│ (BatchNormalization)          │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ re_lu_44 (ReLU)               │ (None, 256, 256, 96)      │               0 │ batch_normalization_65[0]… │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ depthwise_conv2d_22           │ (None, 128, 128, 96)      │             864 │ re_lu_44[0][0]             │
│ (DepthwiseConv2D)             │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ batch_normalization_66        │ (None, 128, 128, 96)      │             384 │ depthwise_conv2d_22[0][0]  │
│ (BatchNormalization)          │                           │               

 Total params: 3,272,685 (12.48 MB)

 Trainable params: 3,240,573 (12.36 MB)

 Non-trainable params: 32,112 (125.44 KB)

In [15]:

vgg16_unet.output_shape

(None, 512, 512, 24)

In [34]:
# pip install graphviz

In [35]:
# pip install pydot

In [36]:
# from IPython.display import SVG
# from tensorflow.keras.utils import model_to_dot

# SVG(model_to_dot(vgg16_unet).create(prog='dot', format='svg'))
# plot_model(vgg16_unet, to_file='vgg16_unet_plot.png', show_shapes=True, show_layer_names=True, expand_nested=True)

In [37]:
from tensorflow.keras.callbacks import LambdaCallback
class DebugBatchCallback(tf.keras.callbacks.Callback):
    def __init__(self):
        super().__init__()  # Initialize the base Callback class
    
    def on_batch_end(self, batch, logs=None):
        # Safely get the 'loss' and 'accuracy' with a default value
        loss = logs.get('loss', 'N/A')
        accuracy = logs.get('accuracy', 'N/A')
        
        # Check if loss and accuracy are floats before formatting
        if isinstance(loss, float):
            loss_str = f"{loss:.4f}"
        else:
            loss_str = str(loss)
            print("h")

        if isinstance(accuracy, float):
            accuracy_str = f"{accuracy:.4f}"
        else:
            accuracy_str = str(accuracy)
        

        # Print batch number, loss, and accuracy (if available)
        print(f"Batch {batch + 1} - Loss: {loss_str}, Accuracy: {accuracy_str}")
        
        # Example of accessing the model (if needed)
        first_layer_weights = self.model.layers[0].get_weights()
        print(f"    - First layer weights shape: {first_layer_weights[0].shape}")
db=DebugBatchCallback()        
def exponential_decay(lr0, s):
    def exponential_decay_fn(epoch):
        return lr0 * 0.1 **(epoch / s)
    return exponential_decay_fn

exponential_decay_fn = exponential_decay(0.0001, 20)

lr_scheduler = LearningRateScheduler(
    exponential_decay_fn,
    verbose=1
)

checkpoint = ModelCheckpoint(
    filepath = 'vgg16_unet_model.keras',
    save_best_only = True, 
#     save_weights_only = False,
    monitor = 'loss', 
    mode = 'auto', 
    verbose = 1
)

earlystop = EarlyStopping(
    monitor = 'loss', 
    min_delta = 0.001, 
    patience = 6, 
    mode = 'auto', 
    verbose = 1,
    restore_best_weights = True
)


csvlogger = CSVLogger(
    filename= "model_training_csv.log",
    separator = ",",
    append = False
)
class timeshistory(tf.keras.callbacks.Callback):
    def __init__(self, file_path="epoch_times.txt"):
        super(timeshistory, self).__init__()
        self.file_path = file_path  # Custom file path for saving the times

    def on_train_begin(self, logs=None):
        self.epoch_times = []  # Initialize a list to store times for each epoch

    def on_epoch_begin(self, epoch, logs=None):
        self.start_time = time.time()  # Record the start time of the epoch

    def on_epoch_end(self, epoch, logs=None):
        epoch_duration = time.time() - self.start_time  # Calculate the duration of the epoch
        self.epoch_times.append(epoch_duration)  # Append to the list

    def on_train_end(self, logs=None):
        # Ensure the directory exists
        os.makedirs(os.path.dirname(self.file_path), exist_ok=True)
        
        # Store the epoch times list in the specified file path
        with open(self.file_path, "w") as file:
            file.write(str(self.epoch_times))
        print(f"Epoch times saved to '{self.file_path}'.")
times_callback = timeshistory(file_path='/home/pranav/DeepLearning/times.txt')
callbacks = [checkpoint,csvlogger, lr_scheduler]
with open("/home/pranav/DeepLearning/new.txt","a") as wr:
    wr.write("The eigth")

In [38]:
print(train_images)
print(train_masks)
os.environ['XLA_FLAGS'] = '--xla_gpu_strict_conv_algorithm_picker=false'
os.environ["CUDA_VISIBLE_DEVICES"] = "1,2,3,4,5,6,7"
print("hh")
print("hhh")      
# print(val_images)
# print(masksval)

/home/pranav/DeepLearning/train_images
/home/pranav/DeepLearning/train_masks
hh
hhh


In [39]:
steps_per_epoch=200
history = vgg16_unet.fit(
    TrainAugmentGenerator(train_images_dir = train_images, train_masks_dir = train_masks,batch_sizess=8, target_sizess = (512, 512)), 
    steps_per_epoch=steps_per_epoch,
    epochs = 100,
    callbacks=callbacks,
    verbose=1
)

Found 1600 images belonging to 1 classes.
Found 1600 images belonging to 1 classes.

Epoch 1: LearningRateScheduler setting learning rate to 0.0001.
Epoch 1/100


I0000 00:00:1731333599.340830  154724 service.cc:148] XLA service 0x7ff4ef017b50 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1731333599.340885  154724 service.cc:156]   StreamExecutor device (0): NVIDIA A100-SXM4-80GB, Compute Capability 8.0
I0000 00:00:1731333599.340890  154724 service.cc:156]   StreamExecutor device (1): NVIDIA A100-SXM4-80GB, Compute Capability 8.0
I0000 00:00:1731333599.340894  154724 service.cc:156]   StreamExecutor device (2): NVIDIA A100-SXM4-80GB, Compute Capability 8.0
I0000 00:00:1731333599.340898  154724 service.cc:156]   StreamExecutor device (3): NVIDIA A100-SXM4-80GB, Compute Capability 8.0
I0000 00:00:1731333599.340901  154724 service.cc:156]   StreamExecutor device (4): NVIDIA A100-SXM4-80GB, Compute Capability 8.0
I0000 00:00:1731333599.340904  154724 service.cc:156]   StreamExecutor device (5): NVIDIA A100-SXM4-80GB, Compute Capability 8.0
I0000 00:00:1731333599.340907  154724 service.cc:156]   S

200/200 ━━━━━━━━━━━━━━━━━━━━ 0s 4s/step - dice_coefficient: 0.3080 - loss: 1.9255
Epoch 1: loss improved from inf to 1.42964, saving model to vgg16_unet_model.keras
200/200 ━━━━━━━━━━━━━━━━━━━━ 1098s 4s/step - dice_coefficient: 0.3086 - loss: 1.9231 - learning_rate: 1.0000e-04

Epoch 2: LearningRateScheduler setting learning rate to 8.912509381337456e-05.
Epoch 2/100
200/200 ━━━━━━━━━━━━━━━━━━━━ 0s 4s/step - dice_coefficient: 0.5905 - loss: 0.9813
Epoch 2: loss improved from 1.42964 to 0.92199, saving model to vgg16_unet_model.keras
200/200 ━━━━━━━━━━━━━━━━━━━━ 889s 4s/step - dice_coefficient: 0.5906 - loss: 0.9810 - learning_rate: 8.9125e-05

Epoch 3: LearningRateScheduler setting learning rate to 7.943282347242815e-05.
Epoch 3/100
200/200 ━━━━━━━━━━━━━━━━━━━━ 0s 4s/step - dice_coefficient: 0.6629 - loss: 0.7885
Epoch 3: loss improved from 0.92199 to 0.77827, saving model to vgg16_unet_model.keras
200/200 ━━━━━━━━━━━━━━━━━━━━ 885s 4s/step - dice_coefficient: 0.6629 - loss: 0.7885 - le

In [40]:
vgg16_unet.save('/home/pranav/DeepLearning/100epochswithouttraining/100epochsunet.keras')

In [41]:
with open('/home/pranav/DeepLearning/100epochswithouttraining/100epochstrainHistoryDict', 'wb') as file_pi:
    pickle.dump(history.history, file_pi)

fig, ax = plt.subplots(1, 3, figsize=(30, 5))
ax = ax.ravel()
metrics = ['Dice Coefficient', 'Loss', 'Learning Rate']

for i, met in enumerate(['dice_coefficient','loss','learning_rate']): 
    if met != 'lr':
        ax[i].plot(history.history[met], 'o-')
        #ax[i].plot(history.history['val_' + met], 'o-')
        ax[i].set_title('{} vs Epochs'.format(metrics[i]), fontsize=16)
        ax[i].set_xlabel('Epochs')
        ax[i].set_ylabel(metrics[i])
        ax[i].set_xticks(np.arange(1,3,2))
        ax[i].legend(['Train', 'Validation'])
        ax[i].xaxis.grid(True, color = "lightgray", linewidth = "0.8", linestyle = "-")
        ax[i].yaxis.grid(True, color = "lightgray", linewidth = "0.8", linestyle = "-")
    else:
        ax[i].plot(history.history[met], 'o-')
        ax[i].set_title('{} vs Epochs'.format(metrics[i]), fontsize=16)
        ax[i].set_xlabel('Epochs')
        ax[i].set_ylabel(metrics[i])
        ax[i].set_xticks(np.arange(1,3,2))
        ax[i].xaxis.grid(True, color = "lightgray", linewidth = "0.8", linestyle = "-")
        ax[i].yaxis.grid(True, color = "lightgray", linewidth = "0.8", linestyle = "-")
        
plt.savefig('model_metrics_plot.png', facecolor= 'w',transparent= False, bbox_inches= 'tight', dpi= 150)

In [42]:
#vgg16_unet.load_weights("./vgg16_unet_model.keras")

In [43]:
# testing_gen = ValAugmentGenerator(val_images_dr = val_images, masksvaldr = masksval, target_sizess = (512, 512))

In [44]:
batch_img,batch_mask = ( "C:/Users/prana/Deep Learning - Segmentation/val_train_images/another/aug_1_000.jpg"  ,"C:/Users/prana/Deep Learning - Segmentation/masksvalimages/another/aug_1_000.png" )
image_path = "C:/Users/prana/Deep Learning - Segmentation/val_train_images/another/aug_1_001.jpg" 
img = cv2.imread(image_path)

# Step 2: Resize and preprocess the image
input_shape = (512, 512)  # Example input size for your model
img_resized = cv2.resize(img, input_shape)
img_normalized = img_resized / 255.0  # Normalize to [0, 1]
img_batch = np.expand_dims(img_normalized, axis=0)  # Add batch dimension

image_path = "C:/Users/prana/Deep Learning - Segmentation/masksvalimages/another/aug_1_001.png"
img = cv2.imread(image_path)

# Step 2: Resize and preprocess the image
input_shape = (512, 512)  # Example input size for your model
img_resized = cv2.resize(img, input_shape)
img_normalized = img_resized / 255.0  # Normalize to [0, 1]
img_mask = np.expand_dims(img_normalized, axis=0)  # Add batch dimension

print(img_batch.shape)

pred_all= vgg16_unet.predict(img_batch)
np.shape(pred_all)
    

fig = plt.figure(figsize=(20,8))

ax1 = fig.add_subplot(1,3,1)
ax1.imshow(np.squeeze(img_batch))
ax1.title.set_text('Original Image')
ax1.grid(False)

ax2 = fig.add_subplot(1,3,2)
ax2.set_title('Ground Truth Mask')
img_mask=np.squeeze(img_mask)
ax2.imshow(onehot_to_rgb(img_mask,id2code))
ax2.grid(False)

ax3 = fig.add_subplot(1,3,3)
ax3.set_title('Predicted Mask')
pred_all=np.squeeze(pred_all)           
ax3.imshow(onehot_to_rgb(pred_all,id2code))
ax3.grid(False)


plt.show()

[ WARN:0@52135.333] global loadsave.cpp:241 findDecoder imread_('C:/Users/prana/Deep Learning - Segmentation/val_train_images/another/aug_1_001.jpg'): can't open/read file: check file path/integrity


error: OpenCV(4.10.0) /io/opencv/modules/imgproc/src/resize.cpp:4152: error: (-215:Assertion failed) !ssize.empty() in function 'resize'


In [45]:
# !zip -r predictions.zip './predictions/'